Approved/multidimensional-arrays.md

```diff
@@ -0,0 +1,657 @@
+- `function Subarray3<'T>(indices : (Int, Int, Int)[], array : 'T[,,]) : 'T[]`
+- `function ConstantArray2<'T>(shape : (Int, Int), value : 'T) : 'T[,]`
+- `function ConstantArray3<'T>(shape : (Int, Int, Int), value : 'T) : 'T[,,]`
+- `function EmptyArray2<'T>() : 'T[,]`
+- `function EmptyArray3<'T>() : 'T[,,]`
+- `function ReducedAlongAxis2<'TInput, 'TOutput>(reduction : ('TInput[] -> 'TOutput), axis : Int, array : 'TInput[,]) : 'TOutput[]`
+- `function ReducedAlongAxis3<'TInput, 'TOutput>(reduction : ('TInput[] -> 'TOutput), axis : Int, array : 'TInput[,,]) : 'TOutput[,]`
+- `function TimesD2(left : Double[,], right : Double[,]) : Double[,]`
+- `function TimesC2(left : Complex[,], right : Complex[,]) : Complex[,]`
+- `function Dot2With2<'T>(plus : (('T, 'T) -> 'T), times : (('T, 'T) -> 'T), (idxLeft : Int, idxRight : Int), left : 'T[,], right : 'T[,]) : 'T[,]`
+- `function Dot3With2<'T>(plus : (('T, 'T) -> 'T), times : (('T, 'T) -> 'T), (idxLeft : Int, idxRight : Int), left : 'T[,,], right : 'T[,]) : 'T[,,]`
+- `function Dot2With3<'T>(plus : (('T, 'T) -> 'T), times : (('T, 'T) -> 'T), (idxLeft : Int, idxRight : Int), left : 'T[,], right : 'T[,,]) : 'T[,,]`
+- `function Dot3With3<'T>(plus : (('T, 'T) -> 'T), times : (('T, 'T) -> 'T), (idxLeft : Int, idxRight : Int), left : 'T[,,], right : 'T[,,]) : 'T[,,,]`
+- `function Reshaped1To2<'T>(array : 'T[], newShape : (Int, Int)) : 'T[,]`
+- `function Reshaped1To3<'T>(array : 'T[], newShape : (Int, Int, Int)) : 'T[,,]`
+- `function Reshaped2To3<'T>(array : 'T[,], newShape : (Int, Int, Int)) : 'T[,,]`
+- `function Reshaped3To3<'T>(array : 'T[,,], newShape : (Int, Int, Int)) : 'T[,,]`
+- `function Reshaped3To2<'T>(array : 'T[,,], newShape : (Int, Int)) : 'T[,]`
+- `function Broadcasted2<'TLeft, 'TRight, 'TOutput>(fn : (('TLeft, 'TRight) -> 'TOutput), left : 'TLeft[,], right : 'TRight[,]) : 'TOutput[,]`
+- `function Broadcasted3<'T0, 'T1, 'T2, 'TOutput>(fn : (('T0, 'T1, 'T2) -> 'TOutput), in0 : 'T0[,], in1 : 'T1[,], in2 : 'T2[,]) : 'TOutput[,]`
+- `function DiagonalMatrix<'T>(diag : 'T[]) : 'T[,]`
+- `function Diagonal<'T>(array : 'T[,]) : 'T[]`
+
+Many of these array functions can be written in Q# itself, while others would require implementation as intrinsics.
+Note that above, we have used suffixes like `2` and `3` to denote ranks of inputs and outputs (or even `2With3` when inputs have different ranks), but these could be eliminated using additional language features, as discussed in "Anticipated Interactions with Future Modifications."
+
+### Examples
+
+Example 1:
+Declaring and indexing into variables of type `Double[,]` and `Int[,,]` using literals.
+
+```qsharp
+let z = #[ // ← #[] marks a two-dimensional array instead of an array of arrays.
+    // Inside #[] delimiters, [] delimiters refer not to array expressions, but
+    // to "rows" of the two-dimensional array literal.
+    [1.0, 0.0],
+    [0.0, -1.0]
+];
+Message($"{z[0, 1]}"); // ← 0.0
+
```

Comment:
> it would save introducing a variation of `[]`

If multidimensional arrays are the default, what is the syntax for jagged array literals? I think you would need a variation of `[]` for jagged arrays still.

Reply:
One option, like in C#, is to not have jagged-array literals (maybe because the are used less often?).

Reply:
C# has literals for both jagged and multidimensional arrays:

```csharp
// Jagged
new int[][]
{
    new int[] { 1, 3, 5, 7, 9 },
    new int[] { 0, 2, 4, 6 },
    new int[] { 11, 22 }
};

// Multidimensional
new int[,]
{
    { 1, 2, 3 },
    { 4, 5, 6 },
    { 7, 8, 9 }
};
```

This corresponds to Q#:

```qsharp
// Jagged
[
    [1, 3, 5, 7, 9],
    [0, 2, 4, 6],
    [11, 22]
];

// Multidimensional
#[
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
];
```

Regardless of whether jagged or multidimensional arrays have the shorter syntax, there needs to be a way to express both of them. Note that Q# doesn't have the `new int[] { 1, 3, 5, 7, 9 }` syntax, and the `new Int[5]` syntax is going away (#48).

Reply:
I meant to say that the jagged-array literal are more like a composite of using the n-d-array literal instead of a special literal.
And I think that's how they avoided creating a new delimiter for that, by having a single n-d-array literal that can be composed into a jagged-array.

But since Q# does not have `new Int[] { 1, 3, 5, 7, 9 }`, then you would need something like:
```
let row1 = [1, 3, 5, 7, 9];
let row2 = [0, 2, 4, 6];
let row3 = [11,22];
let jagged = [row1, row2, row3];
```
which is not optimal...

So an alternative could be to support the _new_ keyword to define array literals:
```
let jagged = 
[
    new [1, 3, 5, 7, 9],
    new [0, 2, 4, 6],
    new [11, 22]
];
```
with the _new_ keyword marking that what comes next is a single object/entry to the array instead of multiple entries.
For consistency, we could always allow the _new_ keyword in front of an array literal and making it implicit/optional, so the following code would also be valid:
```
let nd = new
[
    [1, 2],
    [3, 4],
];
let jagged = new
[
    new [1, 3, 5, 7, 9],
    new [0, 2, 4, 6],
    new [11, 22]
];
```

Reply:
Personally, I find C#'s behavior here rather confusing:

```csharp
int[,][] foo = new int[,][]
{
  {
    new[] {1, 2},
    new[] {3, 4}
  }
};

System.Console.Out.WriteLine($"{foo}");
// prints System.Int32[][,]
```

Regardless, as to your point, @vxfield, I think we could perhaps take some inspiration in making the outer level of nesting (`'T[]`) look different than inner levels:

```qsharp
let rectangular = [
    {1, 2}, // reserve {} inside of [] to indicate levels of rectangular nesting
    {3, 4}
];

let jagged = [ // [] imply "array of," such that jagged arrays remain unchanged
    [1, 3, 5, 7, 9],
    [0, 2, 4, 6],
    [11, 22]
];
```

Alternatively, if we want to avoid reusing `{}`, we could consider other delimiters to indicate nested levels within `[]`:

```qsharp
let rectangular = [
    [| 1, 2 |],
    [| 4, 5 |]
];
```

Regardless of whether we use `{}` or `[| |]`, it's important that inner levels of nesting are not considered expressions in and of themselves. E.g.: `{1, 2, 3}` has no meaning (and is not valid!) outside of `[]`, by contrast to jagged arrays.

Reply:
It may be useful to consolidate all the alternatives side-by-side with 2d, 3d and jagged examples along with their pros and cons.
I like the `[ {} ]` option as it's clean to read but the `{ 1, 2, 3}` may conflict in the future with a potential list literal. Do we have plans for that?
I've attempted to create a table with some options but it's pain to format...

<table>

<tr>
<td width="10%">
alternative
</td>
<td width="30%">
<code>#[ ]</code>
</td>
<td width="30%">
<code>[ { } ]</code>
</td>
<td width="30%">
<code>[| |]</code>
</td>
</tr>

<tr>
<td>
2-d example
</td>
<td>
#[<br>
&nbsp;&nbsp;[ 1, 2 ],<br>
&nbsp;&nbsp;[ 3, 4 ]<br>
]
</td>
<td>
[<br>
&nbsp;&nbsp;{ 1, 2 },<br>
&nbsp;&nbsp;{ 3, 4 }<br>
]
</td>
<td>
[<br>
&nbsp;&nbsp;[| 1, 2 |],<br>
&nbsp;&nbsp;[| 3, 4 |]<br>
]
</td>
</tr>

<tr>
<td>
3-d example
</td>
<td>
##[<br>
&nbsp;&nbsp;[<br>
&nbsp;&nbsp;&nbsp;&nbsp;[ 1, 2 ],<br>
&nbsp;&nbsp;&nbsp;&nbsp;[ 3, 4 ]<br>
&nbsp;&nbsp;],<br>
&nbsp;&nbsp;[<br>
&nbsp;&nbsp;&nbsp;&nbsp;[ 5, 6 ],<br>
&nbsp;&nbsp;&nbsp;&nbsp;[ 7, 8 ]<br>
&nbsp;&nbsp;],<br>
]
</td>
<td>
[<br>
&nbsp;&nbsp;{<br>
&nbsp;&nbsp;&nbsp;&nbsp;{ 1, 2 },<br>
&nbsp;&nbsp;&nbsp;&nbsp;{ 3, 4 }<br>
&nbsp;&nbsp;},<br>
&nbsp;&nbsp;{<br>
&nbsp;&nbsp;&nbsp;&nbsp;{ 5, 6 },<br>
&nbsp;&nbsp;&nbsp;&nbsp;{ 7, 8 }<br>
&nbsp;&nbsp;},<br>
]
</td>
<td>
[<br>
&nbsp;&nbsp;[|<br>
&nbsp;&nbsp;&nbsp;&nbsp;[| 1, 2 |],<br>
&nbsp;&nbsp;&nbsp;&nbsp;[| 3, 4 |]<br>
&nbsp;&nbsp;|],<br>
&nbsp;&nbsp;[|<br>
&nbsp;&nbsp;&nbsp;&nbsp;[| 5, 6 |],<br>
&nbsp;&nbsp;&nbsp;&nbsp;[| 7, 8 |]<br>
&nbsp;&nbsp;|],<br>
]
</td>
</tr>

<tr>
<td>
Pros
</td>
<td>
<ul>
<li>Reuse existing array literals</li>
<li>Inside the array, it's clean to read</li>
</ul>
</td>
<td>
<ul>
<li>Clean to read</li>
</ul>
</td>
<td>
<ul>
<li></li>
</ul>
</td>
</tr>

<tr>
<td>
Cons
</td>
<td>
<ul>
<li>Introduce new <code>#</code> character that could be used for other purposes in the future and potentially creating confusion with other usages</li>
<li>Users need to keep track of how many <code>#</code> are used to understand where the brackets stop being a n-d array and to become 1-d array</li>
</ul>
</td>
<td>
<ul>
<li>May conflict with a future delimiter for list literals</li>
</ul>
</td>
<td>
<ul>
<li>Looks harder to read and noisy with a double-character delimiter</li>
</ul>
</td>
</tr>

</table>

Reply:
Based on the notes I copy-pasted below, there is an objective advantage of having the delimeter/symbol that indicates multi-dimensionality be part of the inner items rather than the containing []: If it is part of the containing [], then by necessity we need to have some way to indicate the dimension/depth for which the restriction of being rectangular applies, whereas having it on the items allows to have a simple rule of [] being the delimiter that "breaks the chain" (see the notes posted in [this comment](https://github.com/microsoft/qsharp-language/pull/49#issuecomment-776404145))